File: leetcode/medium/18_4sum.py
# -----------------------------------------
# My Solution
#
# Time  Complexity: O(n^3)
# Space Complexity: O(n^4)
# -----------------------------------------
import itertools

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []

        all_four_nums = set()
        for index_1, index_2 in itertools.combinations(range(len(nums)), 2):
            for num_1, num_2 in self.all_two_sums(nums, target - nums[index_1] - nums[index_2], set([index_1, index_2])):
                four_num = sorted([nums[index_1], nums[index_2], num_1, num_2])
                all_four_nums.add(tuple(four_num))
        return all_four_nums

    def all_two_sums(self, nums: List[int], target: int, skip_indexes: set[int]) -> List[int]:
        num_count = {}
        for index, num in enumerate(nums):
            if index in skip_indexes:
                continue
            if num not in num_count:
                num_count[num] = 1
            else:
                num_count[num] += 1

        all_two_sums = set()
        for index, num in enumerate(nums):
            if index in skip_indexes:
                continue
            if target - num == num:
                if num_count[num] >= 2:
                    all_two_sums.add((num, num))
            elif target - num in num_count:
                all_two_sums.add((num, target - num))
        return all_two_sums

# -----------------------------------------
# Model Solution
#
# Time  Complexity: O(n^3)
# Space Complexity: O(n^4)
# -----------------------------------------
# Ref: https://leetcode.com/problems/4sum/discuss/292934/python-hashmap-64ms
import collections

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        if len(nums) < 4 or 4 * nums[0] > target or 4 * nums[-1] < target:
            return []
        # use defaultdic for better performance
        diff_indexes_dict = collections.defaultdict(list)
        results = set()

        for index_1 in range(len(nums) - 1):
            for index_2 in range(index_1 + 1, len(nums)):
                diff = target - nums[index_1] - nums[index_2]
                diff_indexes_dict[diff].append((index_1, index_2))

        for index_1 in range(len(nums) - 3):
            for index_2 in range(index_1 + 1, len(nums)):
                total = nums[index_1] + nums[index_2]
                if total in diff_indexes_dict:
                    for indexes in diff_indexes_dict[total]:
                        # Exclude the case indexes[0] or indexes[1] is index_1 or index_2
                        if indexes[0] > index_2:
                            results.add((nums[index_1], nums[index_2], nums[indexes[0]], nums[indexes[1]]))
        return results

File: leetcode/medium/18_4sum.cpp
// -----------------------------------------
// Model Solution: Two Pointers
//
// Time  Complexity: O(n^2)
// Space Complexity: O(log2(n))
// -----------------------------------------
// n := nums.size()
//
// Ref:  https://leetcode.com/problems/4sum/discuss/1341213/C%2B%2BPython-2-solutons-Clean-and-Concise-Follow-up-questions
// Note: Extra Space (without count output as space): O(log2(n)), for sorting space.
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if (nums.size() < 4)
            return {};
        sort(nums.begin(), nums.end());
        if (4 * nums[0] > target || 4 * nums.back() < target)
            return {};

        vector<vector<int>> quadruplets;

        int num_1_index = 0;
        while (num_1_index < nums.size()) {

            int num_2_index = num_1_index + 1;
            while (num_2_index < nums.size()) {
                int num_3_index = num_2_index + 1, num_4_index = nums.size() - 1;
                int target_remain = target - nums[num_1_index] - nums[num_2_index];
                while (num_3_index < num_4_index) {
                    if (nums[num_3_index] + nums[num_4_index] == target_remain) {
                        quadruplets.push_back({nums[num_1_index], nums[num_2_index], nums[num_3_index], nums[num_4_index]});

                        skip_duplicate_numbers(nums, num_3_index);
                        ++num_3_index;
                        skip_duplicate_numbers(nums, num_4_index, -1);
                        --num_4_index;
                    } else if (nums[num_3_index] + nums[num_4_index] > target_remain) {
                        --num_4_index;
                    } else {
                        ++num_3_index;
                    }
                }

                skip_duplicate_numbers(nums, num_2_index);
                ++num_2_index;
            }

            skip_duplicate_numbers(nums, num_1_index);
            ++num_1_index;
        }

        return quadruplets;
    }

private:
    void skip_duplicate_numbers(vector<int>& nums, int& num_index, int move = 1) {
        while (num_index + move >= 0 && num_index + move < nums.size() && nums[num_index] == nums[num_index + move])
            num_index += move;
    }
};

